src/RuleSet/RuleSets.php

```diff
@@ -67,4 +68,45 @@ public static function getSetDefinition(string $name): RuleSetDescriptionInterfa
 
         return $definitions[$name];
     }
+
+    /**
+     * @param class-string<RuleSetDescriptionInterface> $class
+     */
+    public static function registerRuleSet(string $name, string $class): void
+    {
+        $preg = new Preg();
+
+        if (1 !== $preg->match('/^@[a-zA-Z][a-zA-Z0-9_\/\.-]*$/', $name)) {
+            throw new \InvalidArgumentException('RuleSet name must begin with "@" and a letter (a-z, A-Z), and can contain only letters (a-z, A-Z), numbers, underscores, slashes, dots and hyphens.');
+        }
+
+        if (!class_exists($class, true)) {
+            throw new \InvalidArgumentException(sprintf('Class "%s" does not exist.', $class));
+        }
+
+        $preDefinedDefinitions = self::getSetDefinitions();
+
+        if (\array_key_exists($name, $preDefinedDefinitions)) {
+            throw new \InvalidArgumentException(sprintf('Set "%s" is already defined.', $name));
+        }
+
+        if (!\in_array(RuleSetDescriptionInterface::class, class_implements($class), true)) {
+            throw new \InvalidArgumentException(
+                sprintf(
+                    'Class "%s" must be an instance of "%s".',
+                    $class,
+                    RuleSetDescriptionInterface::class
+                )
+            );
+        }
+
+        self::$setDefinitions[$name] = new $class();
+
+        self::sortSetDefinitions();
```

Comment:
why do we need to sort each time we call `registerRuleSet()`?

Reply:
IIRC this logic is copied from somewhere, but not sure. Maybe it's not needed.

Reply:
It's due to sorting only happening in `getSetDefinitions()` if `$setDefinitions` is empty. Seems there's two options:

1. Sort definitions on every call of `getSetDefinitions()`
2. Sort definitions on every call of `registerRuleSet()`

Reply:
3. sort if on `get...()`, but only if previously sorted state got modified

Reply:
True. I left that out as that requires storing or detecting the sorting state, and quickly becomes more painful than other options.

I don't think that sorting while adding is a performance issue, nor is while getting it. To me this is more of a decision if we want to maintain the "correct" state of the `$setDefinitions` at all times or not. I'm not pro premature optimization.

Reply:
After my rework sorting is done at 2 levels:
- _once_ when built-in rulesets are collected (only built-in ones are sorted
- each time when `RuleSets::getSetDefinitions()` is called, because it has to sort built-in rules with custom ones